### ISSUE TITLE
 Tarantool DOUBLE field type and DOUBLE type in SQL

### ISSUE THREAD

**Issue by TarantoolBot** [tarantool/doc#1044]
The DOUBLE field type was added to Tarantool mainly for adding the
DOUBLE type to SQL. Values of this type are stored as MP_DOUBLE in
msgpack. The size of the encoded value is always 9 bytes.

In Lua, only non-integer numbers and CDATA of type DOUBLE can be
inserted in this field. You cannot insert integers of type Lua
NUMBER or CDATA of type int64 or uint64 in this field. The same
rules apply to key in get(), select(), update() and upsert()
methods. It was done this way to avoid unwanted implicit casts
that could affect performance.

It is important to note that you can use the ffi.cast() function
to cast numbers to CDATA of type DOUBLE. An example of this can be
seen below.

Another very important point is that CDATA of type DOUBLE in lua
can be used in arithmetic, but arithmetic for them does not work
correctly. This comes from LuaJIT and most likely will not be
fixed.

Example of usage in Lua:
s = box.schema.space.create('s', {format = {{'d', 'double'}}})
_ = s:create_index('ii')
s:insert({1.1})
ffi = require('ffi')
s:insert({ffi.cast('double', 1)})
s:insert({ffi.cast('double', tonumber('123'))})
s:select(1.1)
s:select({ffi.cast('double', 1)})

In SQL, DOUBLE type behavior is different due to implicit casting.
In a column of type DOUBLE, the number of any supported type can
be inserted. However, it is possible that the number that will be
inserted will be different from that which is inserted due to the
rules for casting to DOUBLE. In addition, after this patch, all
floating point literals will be recognized as DOUBLE. Prior to
that, they were considered as NUMBER.

Example of usage in SQL:
box.execute('CREATE TABLE t (d DOUBLE PRIMARY KEY);')
box.execute('INSERT INTO t VALUES (10), (-2.0), (3.3);')
box.execute('SELECT * FROM t;')
box.execute('SELECT d / 100 FROM t;')
box.execute('SELECT * from t WHERE d < 15;')
box.execute('SELECT * from t WHERE d = 3.3;')
Requested by @ImeevMA in https://github.com/tarantool/tarantool/commit/64745b10c7142b66ab039c7e507be63f1905d465.

**Comment by pgulutzan** [comment]
@ImeevMA: When you write "all floating point literals will be recognized as DOUBLE", is "floating point literal" a literal written with exponential notation, e.g. 1E0 but not 1.0? In Tarantool 2.3.0, pulled from source today, TYPEOF(1E0) is 'number'.

**Comment by ImeevMA** [comment]
@pgulutzan That's my fault. I forgot to fix the typeof() function when adding DOUBLE. I have already created a fix and sent it to review. Still, the returned metadata is correct:
```
tarantool> box.execute('SELECT 1E0')
---
- metadata:
  - name: '1E0'
    type: double
  rows:
  - [1]
...
```

**Comment by pgulutzan** [comment]
@ImeevMA: Thanks for the quick reply. I also asked: When you write "all floating point literals will be recognized as DOUBLE", is "floating point literal" a literal written with exponential notation, e.g. 1E0 but not 1.0?

**Comment by ImeevMA** [comment]
@pgulutzan No, I mean that both 1e0 and 1.0 are of type DOUBLE:
```
tarantool> box.execute('select 1.0')
---
- metadata:
  - name: '1.0'
    type: double
  rows:
  - [1]
...
```

**Comment by pgulutzan** [comment]
@ImeevMA: I regret that you decided that 1.0 is DOUBLE, but I see that nobody else has complained.

Re TYPEOF: you fixed it in version 2.4. Do you intend to fix it in version 2.3? I intend to change the documentation that is marked "2.3.1".

Also, this looks bad: with version 2.4:
create table s (s1 scalar primary key);
insert into s values (cast(1 as number));
insert into s values (cast(2 as unsigned));
select s1,typeof(s1) from s;
The TYPEOF results are not 'number' and 'unsigned'. Is this deliberate and impossible to fix?
